File: ppo/muscle_nn.py
"""
Standalone muscle network module for hierarchical control.

Pure PyTorch implementation with no Ray dependencies.
Compatible with both Python training (ppo_hierarchical.py) and C++ simulation (Environment.cpp).
"""

import torch
import torch.nn as nn
import numpy as np


def weights_init(m):
    """Xavier initialization for linear layers."""
    classname = m.__class__.__name__
    if classname.find('Linear') != -1:
        torch.nn.init.xavier_uniform_(m.weight)
        m.bias.data.zero_()


class MuscleNN(nn.Module):
    """
    Neural network for muscle activation control.

    Maps desired torques and muscle forces to muscle activations.
    Supports both standard and cascading (hierarchical) modes.
    """

    def __init__(self, num_total_muscle_related_dofs, num_dofs, num_muscles, is_cpu=False, is_cascaded=False):
        super(MuscleNN, self).__init__()

        self.num_total_muscle_related_dofs = num_total_muscle_related_dofs
        self.num_dofs = num_dofs  # Exclude Joint Root dof
        self.num_muscles = num_muscles
        self.isCuda = False
        self.isCascaded = is_cascaded

        num_h1 = 256
        num_h2 = 256
        num_h3 = 256

        self.fc = nn.Sequential(
            nn.Linear(num_total_muscle_related_dofs+num_dofs +
                      (num_muscles + 1 if self.isCascaded else 0), num_h1),
            nn.LeakyReLU(0.2, inplace=True),
            nn.Linear(num_h1, num_h2),
            nn.LeakyReLU(0.2, inplace=True),
            nn.Linear(num_h2, num_h3),
            nn.LeakyReLU(0.2, inplace=True),
            nn.Linear(num_h3, num_muscles),
        )

        # Normalization
        self.std_muscle_tau = torch.ones(
            self.num_total_muscle_related_dofs) * 200
        self.std_tau = torch.ones(self.num_dofs) * 200

        if torch.cuda.is_available() and not is_cpu:
            self.isCuda = True
            self.std_tau = self.std_tau.cuda()
            self.std_muscle_tau = self.std_muscle_tau.cuda()
            self.cuda()
            self.device = torch.device('cuda')
        else:
            self.device = torch.device('cpu')

        self.fc.apply(weights_init)

    def forward_with_prev_out_wo_relu(self, muscle_tau, tau, prev_out, weight=1.0):
        muscle_tau = muscle_tau / self.std_muscle_tau
        tau = tau / self.std_tau
        if type(prev_out) == np.ndarray:
            with torch.no_grad():
                prev_out = torch.FloatTensor(prev_out)
                out = prev_out + weight * \
                    self.fc.forward(
                        torch.cat([0.5 * prev_out, weight, muscle_tau, tau], dim=-1))
                return out
        else:
            out = prev_out + weight * \
                self.fc.forward(
                    torch.cat([0.5 * prev_out, weight, muscle_tau, tau], dim=-1))
            return out

    def forward_wo_relu(self, muscle_tau, tau):
        muscle_tau = muscle_tau / self.std_muscle_tau
        tau = tau / self.std_tau
        out = self.fc.forward(torch.cat([muscle_tau, tau], dim=-1))
        return out

    def forward(self, muscle_tau, tau):
        return torch.relu(torch.tanh(self.forward_wo_relu(muscle_tau, tau)))

    def forward_with_prev_out(self, muscle_tau, tau, prev_out, weight=1.0):
        return torch.relu(torch.tanh(self.forward_with_prev_out_wo_relu(muscle_tau, tau, prev_out, weight)))

    def unnormalized_no_grad_forward(self, muscle_tau, tau, prev_out=None, out_np=False, weight=None):
        with torch.no_grad():
            if type(self.std_muscle_tau) == torch.Tensor and type(muscle_tau) != torch.Tensor:
                if self.isCuda:
                    muscle_tau = torch.FloatTensor(muscle_tau).cuda()
                else:
                    muscle_tau = torch.FloatTensor(muscle_tau)

            if type(self.std_tau) == torch.Tensor and type(tau) != torch.Tensor:
                if self.isCuda:
                    tau = torch.FloatTensor(tau).cuda()
                else:
                    tau = torch.FloatTensor(tau)

            if type(weight) != type(None):
                if self.isCuda:
                    weight = torch.FloatTensor([weight]).cuda()
                else:
                    weight = torch.FloatTensor([weight])

            muscle_tau = muscle_tau / self.std_muscle_tau
            tau = tau / self.std_tau

            if type(prev_out) == type(None) and type(weight) == type(None):
                out = self.fc.forward(torch.cat([muscle_tau, tau], dim=-1))
            else:
                if self.isCuda:
                    prev_out = torch.FloatTensor(prev_out).cuda()
                else:
                    prev_out = torch.FloatTensor(prev_out)

                if type(weight) == type(None):
                    print('Weight Error')
                    exit(-1)
                out = self.fc.forward(
                    torch.cat([0.5 * prev_out, weight, muscle_tau, tau], dim=-1))

            if out_np:
                out = out.cpu().numpy()

            return out

    def forward_filter(self, unnormalized_activation):
        return torch.relu(torch.tanh(torch.FloatTensor(unnormalized_activation))).cpu().numpy()

    def load(self, path):
        print('load muscle nn {}'.format(path))
        if torch.cuda.is_available():
            self.load_state_dict(torch.load(path))
        else:
            self.load_state_dict(torch.load(
                path, map_location=torch.device('cpu')))

    def save(self, path):
        print('save muscle nn {}'.format(path))
        torch.save(self.state_dict(), path)

    def get_activation(self, muscle_tau, tau):
        act = self.forward(torch.FloatTensor(muscle_tau.reshape(1, -1)).to(self.device),
                           torch.FloatTensor(tau.reshape(1, -1)).to(self.device))
        return act.cpu().detach().numpy()[0]

    def to(self, *args, **kwargs):
        """Override to() to update self.device"""
        self = super().to(*args, **kwargs)
        # Extract device from args or kwargs
        if args and isinstance(args[0], (torch.device, str)):
            self.device = torch.device(args[0])
        elif 'device' in kwargs:
            self.device = torch.device(kwargs['device'])
        else:
            # Infer device from parameters
            self.device = next(self.parameters()).device
        return self


def generating_muscle_nn(num_total_muscle_related_dofs, num_dofs, num_muscles, is_cpu=False, is_cascaded=False):
    """
    Factory function for C++ pybind11 compatibility.

    Args:
        num_total_muscle_related_dofs: Number of muscle-related DOFs
        num_dofs: Number of actuator DOFs (excluding root)
        num_muscles: Number of muscles
        is_cpu: Force CPU execution (default: False, uses GPU if available)
        is_cascaded: Enable cascading mode for hierarchical control (default: False)

    Returns:
        MuscleNN instance
    """
    return MuscleNN(num_total_muscle_related_dofs, num_dofs, num_muscles, is_cpu, is_cascaded)

File: sim/MuscleNN.cpp
#include "MuscleNN.h"
#include <torch/torch.h>

// Xavier initialization for linear layers
void xavier_init(torch::nn::Linear& layer) {
    torch::nn::init::xavier_uniform_(layer->weight);
    torch::nn::init::zeros_(layer->bias);
}

MuscleNNImpl::MuscleNNImpl(int num_total_muscle_related_dofs, int num_dofs, int num_muscles, bool is_cascaded)
    : num_total_muscle_related_dofs_(num_total_muscle_related_dofs),
      num_dofs_(num_dofs),
      num_muscles_(num_muscles),
      is_cascaded_(is_cascaded),
      device_(torch::kCPU) {

    // Network architecture: 4-layer MLP with LeakyReLU
    const int num_h1 = 256;
    const int num_h2 = 256;
    const int num_h3 = 256;

    // Input dimension depends on cascading mode
    int input_dim = num_total_muscle_related_dofs + num_dofs;
    if (is_cascaded) {
        input_dim += num_muscles + 1;  // prev_out + weight
    }

    // Define layers
    fc1 = register_module("fc1", torch::nn::Linear(input_dim, num_h1));
    fc2 = register_module("fc2", torch::nn::Linear(num_h1, num_h2));
    fc3 = register_module("fc3", torch::nn::Linear(num_h2, num_h3));
    fc4 = register_module("fc4", torch::nn::Linear(num_h3, num_muscles));

    // Initialize normalization parameters (std = 200 for all dimensions)
    std_muscle_tau = torch::ones({num_total_muscle_related_dofs}) * 200.0f;
    std_tau = torch::ones({num_dofs}) * 200.0f;

    // Check for CUDA availability
    if (torch::cuda::is_available()) {
        device_ = torch::kCUDA;
        to(device_);
        std_muscle_tau = std_muscle_tau.to(device_);
        std_tau = std_tau.to(device_);
    }

    // Xavier initialization
    initialize_weights();
}

void MuscleNNImpl::initialize_weights() {
    xavier_init(fc1);
    xavier_init(fc2);
    xavier_init(fc3);
    xavier_init(fc4);
}

torch::Tensor MuscleNNImpl::forward_wo_relu(torch::Tensor muscle_tau, torch::Tensor tau) {
    // Normalize inputs
    muscle_tau = muscle_tau / std_muscle_tau;
    tau = tau / std_tau;

    // Concatenate inputs
    auto x = torch::cat({muscle_tau, tau}, -1);

    // Forward through layers
    x = torch::leaky_relu(fc1->forward(x), 0.2);
    x = torch::leaky_relu(fc2->forward(x), 0.2);
    x = torch::leaky_relu(fc3->forward(x), 0.2);
    x = fc4->forward(x);

    return x;
}

torch::Tensor MuscleNNImpl::forward_with_prev_out_wo_relu(
    torch::Tensor muscle_tau, torch::Tensor tau, torch::Tensor prev_out, double weight) {

    // Normalize inputs
    muscle_tau = muscle_tau / std_muscle_tau;
    tau = tau / std_tau;

    // Create weight tensor
    auto weight_tensor = torch::full({1}, weight, torch::TensorOptions().dtype(torch::kFloat32).device(muscle_tau.device()));

    // Concatenate: [0.5 * prev_out, weight, muscle_tau, tau]
    auto x = torch::cat({0.5 * prev_out, weight_tensor, muscle_tau, tau}, -1);

    // Forward through layers
    x = torch::leaky_relu(fc1->forward(x), 0.2);
    x = torch::leaky_relu(fc2->forward(x), 0.2);
    x = torch::leaky_relu(fc3->forward(x), 0.2);
    auto out = fc4->forward(x);

    // Add residual connection: out = prev_out + weight * network_output
    return prev_out + weight * out;
}

torch::Tensor MuscleNNImpl::forward(torch::Tensor muscle_tau, torch::Tensor tau) {
    return torch::relu(torch::tanh(forward_wo_relu(muscle_tau, tau)));
}

torch::Tensor MuscleNNImpl::forward_with_prev_out(
    torch::Tensor muscle_tau, torch::Tensor tau, torch::Tensor prev_out, double weight) {
    return torch::relu(torch::tanh(forward_with_prev_out_wo_relu(muscle_tau, tau, prev_out, weight)));
}

Eigen::VectorXf MuscleNNImpl::unnormalized_no_grad_forward(
    const Eigen::VectorXd& muscle_tau,
    const Eigen::VectorXd& tau,
    const Eigen::VectorXf* prev_out,
    double weight) {

    torch::NoGradGuard no_grad;  // Disable gradient computation

    // Convert Eigen (double) → torch::Tensor (float32)
    auto muscle_tau_tensor = torch::from_blob(
        const_cast<double*>(muscle_tau.data()),
        {muscle_tau.size()},
        torch::kFloat64
    ).to(torch::kFloat32).to(device_);

    auto tau_tensor = torch::from_blob(
        const_cast<double*>(tau.data()),
        {tau.size()},
        torch::kFloat64
    ).to(torch::kFloat32).to(device_);

    torch::Tensor out;

    if (prev_out == nullptr) {
        // Standard mode (non-cascading)
        out = forward_wo_relu(muscle_tau_tensor, tau_tensor);
    } else {
        // Cascading mode with previous output
        auto prev_out_tensor = torch::from_blob(
            const_cast<float*>(prev_out->data()),
            {prev_out->size()},
            torch::kFloat32
        ).to(device_);

        out = forward_with_prev_out_wo_relu(muscle_tau_tensor, tau_tensor, prev_out_tensor, weight);
    }

    // Convert result back to Eigen (float32)
    out = out.to(torch::kCPU);
    Eigen::VectorXf result(out.size(0));
    std::memcpy(result.data(), out.data_ptr<float>(), out.size(0) * sizeof(float));

    return result;
}

Eigen::VectorXf MuscleNNImpl::forward_filter(const Eigen::VectorXf& unnormalized) {
    torch::NoGradGuard no_grad;

    // Convert Eigen → torch::Tensor
    auto tensor = torch::from_blob(
        const_cast<float*>(unnormalized.data()),
        {unnormalized.size()},
        torch::kFloat32
    ).clone();  // Clone to avoid modifying original

    // Apply activation: relu(tanh(x))
    auto filtered = torch::relu(torch::tanh(tensor));

    // Convert back to Eigen
    Eigen::VectorXf result(filtered.size(0));
    std::memcpy(result.data(), filtered.data_ptr<float>(), filtered.size(0) * sizeof(float));

    return result;
}

void MuscleNNImpl::load_state_dict(const std::unordered_map<std::string, torch::Tensor>& state_dict) {
    // Load each parameter by name
    for (const auto& pair : state_dict) {
        const std::string& name = pair.first;
        const torch::Tensor& param = pair.second;

        // Map Python parameter names to C++ module names
        torch::Tensor* target_param = nullptr;

        if (name == "fc.0.weight") target_param = &fc1->weight;
        else if (name == "fc.0.bias") target_param = &fc1->bias;
        else if (name == "fc.2.weight") target_param = &fc2->weight;
        else if (name == "fc.2.bias") target_param = &fc2->bias;
        else if (name == "fc.4.weight") target_param = &fc3->weight;
        else if (name == "fc.4.bias") target_param = &fc3->bias;
        else if (name == "fc.6.weight") target_param = &fc4->weight;
        else if (name == "fc.6.bias") target_param = &fc4->bias;

        if (target_param != nullptr) {
            // Copy parameter values
            target_param->data().copy_(param.to(device_));
        }
    }
}
